# ENVS 117 Week 2 Tuesday: In-class Exercises

Exercises T1 to T5 (Chapters 8 to 10). Work in pairs. T1 to T4 are the core exercises (about 36 min); T5 is a stretch for fast finishers.

**How to use this notebook:** run each data cell first, then write your answer in the `# Your code here` cell under each question. All data are synthetic teaching values.

Name(s): Asha LaManque

## T1. Urban heat and building materials (Ch 8)

Surface temperatures at 3 pm on a September heat day (°F).

In [14]:
asphalt_f = 142.0      # parking lot
concrete_f = 118.5     # sidewalk
cool_roof_f = 96.8     # white reflective roof
green_roof_f = 88.3    # vegetated roof
air_temp_f = 94        # air temperature at the same time

**1.** Print the `type()` of `asphalt_f` and `air_temp_f`. Why is one an `int`? (Answer in a comment.)

In [15]:
# Your code here
print(type(asphalt_f), type(air_temp_f))

#air_temp_f is an int, while asphalt_f is a float.

<class 'float'> <class 'int'>


**2.** Convert asphalt and green roof temperatures to Celsius and compute the difference (1 decimal).

In [16]:
# Your code here
asphalt_c = round((asphalt_f - 32) * (5/9),1)
print(asphalt_c)
green_roof_c = round((green_roof_f - 32) * (5/9),1)
print(green_roof_c)

61.1
31.3


**3.** Boolean `cool_roof_above_air`: is the cool roof still hotter than the air?

In [17]:
# Your code here
cool_roof_above_air  = cool_roof_f > air_temp_f
print(cool_roof_above_air)

True


**4.** A 2,500 m² roof is retrofitted so 40% of it is green roof. Green area, its type, and the total for 12 identical buildings.

In [18]:
# Your code here
green_area = 2500 * 0.40
print(f"Green Area:", green_area)
print(f"Green Area Type:", type(green_area))
print(f"Green Area Total (12 buildings):",green_area * 12)

Green Area: 1000.0
Green Area Type: <class 'float'>
Green Area Total (12 buildings): 12000.0


**5.** Print one summary line with an f-string, e.g. `Asphalt 61.1 C, green roof 31.3 C, difference 29.8 C`.
*Discussion:* which material choices would you prioritize on a campus master plan, and why?

In [19]:
# Your code here
print(f"Asphalt", asphalt_c, "C, green roof", green_roof_c, "C, difference", asphalt_c - green_roof_c, "C")

#Discussion: I would prioritize a green roof because it has a lower temperature compared to asphalt by 31.3 C. 
#This would reduce additional heat produced and help keep the campus cooler.

Asphalt 61.1 C, green roof 31.3 C, difference 29.8 C


## T2. Creek flow monitoring (Ch 9: tuples, lists, dictionaries)

Daily mean flow on Saratoga Creek (cubic feet per second); a storm hits on day 4.

In [20]:
gauge_location = (37.2710, -122.0330)
daily_flow_cfs = [4.2, 3.9, 3.7, 12.8, 25.4, 9.6, 6.1]

**1.** Unpack `gauge_location` into `lat` and `lon`, then try to change the latitude. Read the error: why does a tuple make sense here?

In [21]:
# Your code here
lat, long = gauge_location[0], gauge_location[1]
print(lat, long)

#Trying to change lat:
#gauge_location[0] = 35.0
#A tuple makes sense because the latitude and longitude for a place should not change. Thus, having an error that describes this issue and prevents it from changing is also helpful.

37.271 -122.033


**2.** Append today's reading, `5.3`, and print the number of days recorded.

In [22]:
# Your code here
daily_flow_cfs.append(5.3)
print(f"Number of days recorded:", len(daily_flow_cfs))

Number of days recorded: 8


**3.** Last three days (slice), peak, minimum and mean (2 decimals).

In [23]:
# Your code here
print(daily_flow_cfs)
#Last 3 days
print(daily_flow_cfs[-3: ])

#peak
print(max(daily_flow_cfs))
#min
print(min(daily_flow_cfs))
#mean
print(sum(daily_flow_cfs)/len(daily_flow_cfs))


[4.2, 3.9, 3.7, 12.8, 25.4, 9.6, 6.1, 5.3]
[9.6, 6.1, 5.3]
25.4
3.7
8.875


**4.** On which day (1-based) did the peak occur? Hint: `daily_flow_cfs.index(value)` returns a position.

In [24]:
# Your code here
peak_day = daily_flow_cfs.index(max(daily_flow_cfs)) + 1
print(f"Peak day:", peak_day)

#Answer: Day 5

Peak day: 5


**5.** Build a dictionary `gauge` with keys `gauge_id`, `stream`, `location`, `flow_cfs`, `units`. Add a new key `peak_cfs`. Use `.get()` to look up `drainage_area_km2` with a default of `"unknown"`.

In [28]:
# Your code here
gauge = ["gauge_id", "strem", "location", "flow_cfs", "units"]
gauge = dict.fromkeys(gauge)
# 2. Add a new key 'peak_cfs' using assignment
gauge["peak_cfs"] = 320.0

# 3. Use .get() to look up 'drainage_area_km2' with a default of "unknown"
drainage_area = gauge.get("drainage_area_km2", "unknown")

print(gauge)
print("Drainage Area:", drainage_area)

{'gauge_id': None, 'strem': None, 'location': None, 'flow_cfs': None, 'units': None, 'peak_cfs': 320.0}
Drainage Area: unknown


## T3. Bird survey comparison (Ch 9: sets)

One entry per sighting, so species repeat.

In [29]:
campus_birds = ["Anna's Hummingbird", "American Crow", "Western Bluebird", "Lesser Goldfinch",
                "American Crow", "Black Phoebe", "Anna's Hummingbird", "California Scrub-Jay"]
ulistac_birds = ["Black Phoebe", "Red-tailed Hawk", "Anna's Hummingbird", "Great Egret",
                 "California Scrub-Jay", "Bushtit", "Black Phoebe", "White-crowned Sparrow"]

**1.** Convert each list to a set. Compare the number of sightings with the number of unique species (species richness) at each site.

In [30]:
# Your code here
campus_birds_sight = len(campus_birds)
campus_birds = set(campus_birds)
campus_birds_unique = len(campus_birds)

print(f"Campus Bird Sightings:" ,campus_birds_sight, ", Campus Birds Unique Species:", campus_birds_unique)

ulistac_birds_sight = len(ulistac_birds)
ulistac_birds = set(ulistac_birds)
ulistac_birds_unique = len(ulistac_birds)

print(f"Ulistac Bird Sightings:" ,ulistac_birds_sight, ", Ulistac Birds Unique Species:", ulistac_birds_unique)




Campus Bird Sightings: 8 , Campus Birds Unique Species: 6
Ulistac Bird Sightings: 8 , Ulistac Birds Unique Species: 7


**2.** Species at both sites, only on campus, only at Ulistac.

In [34]:
# Your code here
print(campus_birds & ulistac_birds)
print(campus_birds - ulistac_birds)
print(ulistac_birds - campus_birds)

{'Black Phoebe', "Anna's Hummingbird", 'California Scrub-Jay'}
{'American Crow', 'Lesser Goldfinch', 'Western Bluebird'}
{'Bushtit', 'Great Egret', 'White-crowned Sparrow', 'Red-tailed Hawk'}


**3.** Combined species richness.

In [35]:
# Your code here
print(campus_birds | ulistac_birds)

{'Bushtit', 'Red-tailed Hawk', "Anna's Hummingbird", 'Great Egret', 'American Crow', 'Black Phoebe', 'Lesser Goldfinch', 'California Scrub-Jay', 'White-crowned Sparrow', 'Western Bluebird'}


**4.** Was a Western Bluebird seen at Ulistac? Answer with `in`.

*Discussion:* why would a list be the wrong structure for "species present", but the right one for "sightings"?

In [36]:
# Your code here
print("Western Bluebird" in ulistac_birds)


False


## T4. Parsing a power grid record (Ch 10)

One raw line from a (synthetic) substation log in Santa Clara.

In [61]:
raw = "  sub-07 | 2026-09-15 18:00 | load=48.6 MW | capacity=55 MW  "
lat, lon = 37.3541, -121.9552

**1.** Strip and `split("|")`. What is still wrong with each piece?

In [62]:
# Your code here
divided = raw.split("|")
print(divided)
#There is an extra space. Also, it could be formatted better. 

['  sub-07 ', ' 2026-09-15 18:00 ', ' load=48.6 MW ', ' capacity=55 MW  ']


**2.** Clean uppercase ID, date and time.

In [80]:
# Your code here
raw = raw.upper()
divided = raw.split(" | ")
print(divided)
date_part, time_part = divided[1].split(' ')
divided[0] = divided[0].replace(" ", "")
divided[-1] = divided[-1].replace("W ", "W")

['  SUB-07', '2026-09-15 18:00', 'LOAD=48.6 MW', 'CAPACITY=55 MW  ']


**3.** Numeric load and capacity (floats) plus units.

In [131]:
# Your code here
print(divided[2][divided[2].index("=") + 1:divided[2].index(" ")], "MW")
print(divided[3][divided[3].index("=") + 1: divided[3].index(" ")], "MW")


48.6 MW
48.6
55 MW


**4.** Print: `Substation SUB-07 | 2026-09-15 at 18:00 | load 48.6 of 55 MW (88.4%)` (use `:.1%`).

In [145]:
# Your code here
load = divided[2][divided[2].index("=") + 1:divided[2].index(" ")]
capacity = divided[3][divided[3].index("=") + 1: divided[3].index(" ")]
percent = round(float(load)/float(capacity) * 100, 1)
percent_string = "(" + str(percent) + "%)"
print(f"Substation", divided[0], "|", date_part, "at", time_part, "|", "load", str(load), "of", str(capacity),"MW", percent_string)

Substation SUB-07 | 2026-09-15 at 18:00 | load 48.6 of 55 MW (88.4%)


**5.** WKT point (longitude first).

In [90]:
# Your code here
point = (lon, lat)

**6.** Build an output file name like `sub-07_20260915_load.csv` using `lower()` and `replace()`.

*Discussion:* why is 6 pm the peak hour for many grids in a heat wave?

In [92]:
# Your code here
output = divided[0].lower() + "_" + date_part.replace("-","") + "_" + "load.csv"
print(output)

sub-07_20260915_load.csv


## T5 (stretch). Reservoir storage report (nested dictionaries)

In [119]:
reservoirs = {
    "Reservoir A": {"capacity_af": 90000, "storage_af": 31500},
    "Reservoir B": {"capacity_af": 22500, "storage_af": 17800},
    "Reservoir C": {"capacity_af": 4800, "storage_af": 1150},
}

**1.** Print each reservoir's percent full using `:.1%`, and Reservoir A with thousands separators: `Reservoir A: 31,500 of 90,000 af (35.0% full)`.

In [120]:
# Your code here
for key in reservoirs.keys():
    print(round(((reservoirs[key].get("storage_af") / reservoirs[key].get("capacity_af")) *100),1))


35.0
79.1
24.0


**2.** Compute the percent full of the whole system (total storage / total capacity).

In [146]:
# Your code here
total_storage = 0
total_capacity = 0
for key in reservoirs.keys():
    total_storage += reservoirs[key].get("storage_af")
    total_capacity += reservoirs[key].get("capacity_af")
print(total_storage/total_capacity * 100)

43.22250639386189


**3.** A storm adds 250 acre-feet to Reservoir C. Update the nested value and print its dictionary.

In [122]:
# Your code here
reservoirs["Reservoir C"]["storage_af"] += 250
print(reservoirs.get("Reservoir C"))

{'capacity_af': 4800, 'storage_af': 1400}


**4.** Reflection: you just wrote nearly the same line three times. What tool would remove the repetition? (Answer in a comment.)

In [123]:
# Your code here
#A for loop would help loop through and help with repetition.